### ISSUE TITLE
Logback MDC inherited in child threads

### ISSUE THREAD

**Issue by francesco-monorchio** [micronaut-projects/micronaut-core#8309]
### Expected Behavior

I want to inherit mdc key value pairs in the child threads but some threads do not work as expected.

I have some HttpServerFilters that set some values in the MDC. The Controllers, instead, are executed with @ ExecuteOn(TaskExecutors.IO) on a different thread pool. For that, I have customized the implementation of the thread factory but some of the threads do not work as expected.

### Actual Behaviour

Some of the IO threads does not inherit MDC values.

### Steps To Reproduce

Run './gradlew run'
With postman, call GET /run-indefinitely

In the logs you should see inherited MDC values.

### Environment Information

OpenSUSE Tumbleweed
Java 17.0.3

### Example Application

https://github.com/francesco-monorchio/micronaut-samples/tree/inheritable-mdc

### Version

3.7.3

**Comment by yawkat** [comment]
it's a thread pool, the thread factory isn't called for every request. but we should provide a way to do this

**Comment by yawkat** [comment]
duplicate of #7914

**Comment by jkebinger** [comment]
> it's a thread pool, the thread factory isn't called for every request. but we should provide a way to do this

@yawkat Is there work planned to provide a way to do this? In principle one just needs to be able to hook the before and after methods on ThreadPoolExecutor to copy thread locals around. Thank you!

**Comment by yawkat** [comment]
it's not planned atm. i doubt there is a clean way, we don't know which thread locals to copy and which should stay.

imo the proper way to do this in the future will be with scoped values from jep 429

**Comment by jkebinger** [comment]
Thank you for the pointer to jep 429. Do you think micronaut will wait on widespread usage of java 20+ to look more at the issue of thread to thread state transfer?

At my last place of work, we implemented a ThreadLocalFollower system where components that wanted data copied would register using java SPI/serviceloader and a common executor implementation that would call the registered ThreadLocalFollower(s) to both set and later clear the thread context they managed, which worked quite well.

Spring seems to have an approach with ThreadLocalTargetSource but I don't completely grok how it may clean up itself when work stops using a pooled thread.


Is it possible and/or advised to use the `io.micronaut.tracing.instrument.util` factories to copy thread local state around in this manner the way the [MdcInstrumenter](https://github.com/micronaut-projects/micronaut-tracing/blob/master/tracing-core/src/main/java/io/micronaut/tracing/instrument/util/MdcInstrumenter.java) does?

**Comment by yawkat** [comment]
@dstepanov is looking at more general propagation improvements. we have a problem with TLs in general because we use reactive / async code all the time so TLs tend to get lost.

We will have to figure out something that does not rely on loom because mn 4 will be based on java 17. however imo in the long term i think jep 429 is the better way. the main point of that jep is fixing propagation, since with loom, child threads will be common. so ideally at some point in mn 4.x, you will be able to write an application where MDC and everything similar is handled using jep 429 and propagated seamlessly (though we cant require its use because of java 17 compat).

**Comment by graemerocher** [comment]
with reactive stacks it is simply not possible to use thread local implicitly unfortunately. Even Spring you have to include manual integration with Spring Web Reactive. See https://simonbasle.github.io/2018/02/contextual-logging-with-reactor-context-and-mdc/

**Comment by yawkat** [comment]
Yes the instrumenter is how we "fix" MDC right now. however it has had bugs before and it's not foolproof. hence looking forward to @dstepanov's work on this :)

**Comment by graemerocher** [comment]
will be probably the third time @dstepanov has re-written instrumentation
